openedx/core/djangoapps/oauth_dispatch/views.py

```diff
@@ -75,6 +75,39 @@ def _get_client_id(self, request):
             return request.POST.get('client_id')
 
 
+def _get_token_type(request):
+    """
+    Get the token_type for the request.
+
+    - Respects the HTTP_X_TOKEN_TYPE header if the token_type parameter is not supplied.
+    - Adds `oauth_token_type` custom attribute for monitoring.
+    """
+    default_token_type = request.META.get('HTTP_X_TOKEN_TYPE', 'no_token_type_supplied')
+    token_type = request.POST.get('token_type', default_token_type).lower()
+    monitoring_utils.set_custom_attribute('oauth_token_type', token_type)
+    return token_type
+
+
+def _get_jwt_dict_from_opaque_access_token_dict(opaque_token_dict, oauth_adapter):
+    """
+    Returns a JWT token dict from the provided opaque access token dict.
+
+    Creates the new JWT, and then overrides various values in a copy of the
+        token dict with the JWT specific values.
+    """
+    jwt_dict = opaque_token_dict.copy()
+    # TODO: It would be safer if create_jwt_from_token returned this
+    #   dict directly, so it would not be possible for the dict and JWT
+    #   to get out of sync, but that is a larger refactor to think through.
+    jwt = create_jwt_from_token(jwt_dict, oauth_adapter)
```

Comment:
**Note to Reviewer:** Ideally, I'd like there to be a new method named something like `replace_token_dict_with_jwt_token_dict` (or something like that), that lives in the module with `create_jwt_from_token`, and both creates the JWT **and** replaces the appropriate token dict fields. Then, the method `create_jwt_from_token` could just call `replace_token_dict_with_jwt_token_dict` and return `jwt_token_dict['access_token']`, with a warning that it should only be used in cases where the updated token dict is not needed. We could then search uses of `create_jwt_from_token` and replace it as needed.

However, this would also require some unit test refactoring, and I want to limit work done here. @jawad-khan: Is this a minor refactor you could pick up after this PR merges?

Reply:
[inform] I decided to do this refactor in this PR: https://github.com/openedx/edx-platform/pull/30485